File: src/vector.ts
export class Vector2D{
    x: number
    y: number

    constructor(x: number, y: number){
        this.x = x
        this.y = y
    }

    translate(b: Vector2D): Vector2D { return Vector2D.translate(this, b) }
    subtract(a: Vector2D): Vector2D { return Vector2D.difference(a, this) }
    dot(b: Vector2D): number { return Vector2D.dot(this, b) }
    scale(scale: number): Vector2D { return Vector2D.scale(this, scale) }
    scaleV(scale: Vector2D): Vector2D { return Vector2D.scaleV(this, scale) }

    length(): number{
        return Math.sqrt(this.x*this.x + this.y*this.y)
    }

    direction(length: number = this.length()): Vector2D{
        if(length == 0) return new Vector2D(0,0)
        return new Vector2D(this.x/length, this.y/length)
    }

    static translate(a: Vector2D, b: Vector2D): Vector2D{ 
        return new Vector2D(a.x + b.x, a.y + b.y)
    }
    static difference(a: Vector2D, b: Vector2D): Vector2D{ 
        return new Vector2D(b.x - a.x, b.y - a.y)
    }
    static dot(a: Vector2D, b: Vector2D): number{ 
        return a.x * b.x + a.y * b.y
    }
    static scale(a: Vector2D, scale: number): Vector2D{ 
        return new Vector2D(a.x * scale, a.y * scale)
    }
    static scaleV(a: Vector2D, scale: Vector2D): Vector2D{ 
        return new Vector2D(a.x * scale.x, a.y * scale.y)
    }
}

export class Vector3D{
    x: number
    y: number
    z: number

    constructor(x: number, y: number, z: number){
        this.x = x
        this.y = y
        this.z = z
    }

    translate(b: Vector3D): Vector3D { return Vector3D.translate(this, b) }
    subtract(a: Vector3D): Vector3D { return Vector3D.difference(a, this) }
    dot(b: Vector3D): number { return Vector3D.dot(this, b) }
    scale(scale: number): Vector3D { return Vector3D.scale(this, scale) }
    scaleV(scale: Vector3D): Vector3D { return Vector3D.scaleV(this, scale) }

    length(): number{
        return Math.sqrt(this.x*this.x + this.y*this.y + this.z*this.z)
    }

    direction(length: number = this.length()): Vector3D{
        if(length == 0) return new Vector3D(0,0,0)
        return new Vector3D(this.x/length, this.y/length, this.z/length)
    }

    static translate(a: Vector3D, b: Vector3D): Vector3D{ 
        return new Vector3D(a.x + b.x, a.y + b.y, a.z + b.z)
    }
    static difference(a: Vector3D, b: Vector3D): Vector3D{ 
        return new Vector3D(b.x - a.x, b.y - a.y, b.z - a.z)
    }
    static dot(a: Vector3D, b: Vector3D): number{ 
        return a.x * b.x + a.y * b.y + a.z * b.z
    }
    static scale(a: Vector3D, scale: number): Vector3D{ 
        return new Vector3D(a.x * scale, a.y * scale, a.z * scale)
    }
    static scaleV(a: Vector3D, scale: Vector3D): Vector3D{ 
        return new Vector3D(a.x * scale.x, a.y * scale.y, a.z * scale.z)
    }
}

File: src/Vector.kt

import kotlin.math.sqrt

typealias Color = Vector4

interface Vector {
    fun components(): List<Float>
    fun length(): Float
    @JsName("dot") fun dot(other: Vector): Float
    fun calculateLength() = sqrt(this.dot(this))
}

class InvalidVectorOperation(dimensions: Int): IllegalArgumentException("This operation is invalid, may only operate on Vector$dimensions")

data class Vector2(val x: Float, val y: Float): Vector{
    private var _length: Float? = null
    @JsName("plus") operator fun plus(other: Vector2) = Vector2(x = x + other.x, y = y + other.y)
    @JsName("minus") operator fun minus(other: Vector2) = Vector2(x = x - other.x, y = y - other.y)
    @JsName("scale") operator fun times(scale: Vector2) = Vector2(x = x * scale.x, y = y * scale.y)
    @JsName("scaleUniformly") operator fun times(scale: Float) = Vector2(x = x * scale, y = y * scale)
    override fun dot(other: Vector) = when(other){ is Vector2 -> x * other.x + y * other.y; else -> throw InvalidVectorOperation(2) }
    override fun length(): Float {
        if(_length == null) _length = calculateLength()
        return _length!!
    }
    override fun components() = listOf(x, y)
}

data class Vector3(val x: Float, val y: Float, val z: Float): Vector {
    private var _length: Float? = null
    @JsName("plus") operator fun plus(other: Vector3) = Vector3(x = x + other.x, y = y + other.y, z = z + other.z)
    @JsName("minus") operator fun minus(other: Vector3) = Vector3(x = x - other.x, y = y - other.y, z = z - other.z)
    @JsName("scale") operator fun times(scale: Vector3) = Vector3(x = x * scale.x, y = y * scale.y, z = z * scale.z)
    @JsName("scaleUniformly") operator fun times(scale: Float) = Vector3(x = x * scale, y = y * scale, z = z * scale)
    override fun dot(other: Vector) = when(other){ is Vector3 -> x * other.x + y * other.y + z * other.z; else -> throw InvalidVectorOperation(3) }
    override fun length(): Float {
        if (_length == null) _length = calculateLength()
        return _length!!
    }
    override fun components() = listOf(x, y, z)
}

data class Vector4(val r: Float, val g: Float, val b: Float, val a: Float): Vector {
    private var _length: Float? = null
    @JsName("plus") operator fun plus(other: Vector4) = Vector4(r = r + other.r, g = g + other.g, b = b + other.b, a = a + other.a)
    @JsName("minus") operator fun minus(other: Vector4) = Vector4(r = r - other.r, g = g - other.g, b = b - other.b, a = a - other.a)
    @JsName("scale") operator fun times(scale: Vector4) = Vector4(r = r * scale.r, g = g * scale.g, b = b * scale.b, a = a * scale.a)
    @JsName("scaleUniformly") operator fun times(scale: Float) = Vector4(r = r * scale, g = g * scale, b = b * scale, a = a * scale)
    override fun dot(other: Vector) = when(other){ is Vector4 -> r * other.r + g * other.g + b * other.b + a * other.a; else -> throw InvalidVectorOperation(4) }
    override fun length(): Float {
        if (_length == null) _length = calculateLength()
        return _length!!
    }
    override fun components() = listOf(r, g, b, a)
}
